### PR TITLE
chore(eslint): fine-tune eslint and fix most errors

### PR DESCRIPTION
**Pull request by nchilada**
In service of #2657, the monorepo now has a pre-commit hook that runs eslint on staged files, but there's still a bit more work to be done.

In this PR:

* Special handling many common files that would currently fail the lint check if/when they are changed and staged for commit:
  * JSON files
  * JS files containing JSON module imports (`import ... assert {type: 'json'}`
  * Markdown files, with JS linting for the embedded code examples
* Make a distinction between frontend files (which may use Web APIs, React/Next globals, JSX syntax, etc.) and backend files (which may use Node APIs)
* Ignore some files: `dist`, build scripts, etc.
* Add a linting job to CI, for all files changed in a given pull request. (We already have a job to assess lint on all files, but it's currently failing and seems likely to fail at times in the future too.) 
* Remove the legacy `.eslintrc` files that had been specified in a subset of the codebase.

Future work?

* Get all files to pass eslint and prettier, to satisfy the all-files check in CI and to reduce churn in future pull requests

### CODE REVIEW

File: .eslintrc.cjs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+  'sites/*/themes/**',
+]
+const frontendFiles = [
+  '**/components/**',
+  '**/hooks/**',
+  '**/pages/**',
+  '**/page-templates/**',
+  'packages/i18n/**/*.md/*.js',
+]
+const isomorphicFiles = ['packages/core/**', 'plugins/**', 'tests/designs/**', 'tests/plugins/**']
```
Comment (nchilada) [inline .eslintrc.cjs:20]: Based on the [distinction](https://freesewing.dev/guides/plugins/types-of-plugins) between build-time and run-time plugins, I gathered that plugin code should be able to run in either Node.js or a web browser, and therefore ought to restrict itself to the intersection of Node builtins and browser builtins. (And I'm guessing the same is true for some of the core code.) Ironically, however, the only plugin that [seems to use](https://github.com/freesewing/freesewing/blob/a44c23c270546a2229da2a583ea33405e7bcaaab/plugins/plugin-timing/src/index.mjs#L4-L11) any such builtins actually uses items from the _union_ of Node builtins and browser builtins, and will therefore need an `eslint-ignore` directive in order to comply with the lint rules. So maybe I'm over-engineering here.

Comment (nchilada) [inline .eslintrc.cjs:20]: Based on the [distinction](https://freesewing.dev/guides/plugins/types-of-plugins) between build-time and run-time plugins, I gathered that plugin code should be able to run in either Node.js or a web browser and therefore ought to restrict itself to the _intersection_ of Node builtins and browser builtins. (I'm guessing the same is true for some of the core code, but I'm even less sure about that.)

Ironically, however, the only plugin that [seems to use](https://github.com/freesewing/freesewing/blob/a44c23c270546a2229da2a583ea33405e7bcaaab/plugins/plugin-timing/src/index.mjs#L4-L11) any such builtins actually uses items from the _union_ of Node builtins and browser builtins, and will therefore need an `eslint-ignore` directive in order to comply with the lint rules.

So maybe I'm over-engineering here.

File: .eslintrc.cjs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        'no-console': 'off',
+      },
+    },
+    {
+      files: frontendFiles,
+      excludedFiles: nodeFiles,
+      extends: ['next/core-web-vitals'],
+      env: {
+        // We can be stricter than 'next/core-web-vitals' is
+        node: false,
```
Comment (nchilada) [inline .eslintrc.cjs:56]: The `next/core-web-vitals` eslint config, which is recommended for use throughout a Next.js project, specifies `node: true, browser: true` and therefore technically allows JavaScript to access Node and browser builtins, even where that might not make sense. I'm trying to be stricter than that, here, but I may be over-engineering again.

File: .eslintrc.cjs [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    {
+      files: isomorphicFiles,
+      excludedFiles: [].concat(nodeFiles, frontendFiles),
+      env: {
+        'shared-node-browser': true,
+      },
+    },
+    // Additional globals for JavaScript files that happen to be CommonJS.
+    // Only allowed in *.cjs files, not *.js files, because we probably want to move towards
+    // a `"type": "module"` future where any CommonJS files would have to have .cjs extensions.
```
Comment (nchilada) [inline .eslintrc.cjs:68]: More background [here](https://nodejs.org/api/packages.html#determining-module-system), though it doesn't directly explain why one might prefer ESM.

File: .eslintrc.cjs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+      },
+    },
+    {
+      files: isomorphicFiles,
+      excludedFiles: [].concat(nodeFiles, frontendFiles),
+      env: {
+        'shared-node-browser': true,
+      },
+    },
+    // Additional globals for JavaScript files that happen to be CommonJS.
```
Comment (nchilada) [inline .eslintrc.cjs:66]: `require`,` `module`, `exports`

File: .eslintrc.cjs [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+  },
+  // Required when using experimental EcmaScript features
+  parser: '@babel/eslint-parser',
+  parserOptions: {
+    ecmaVersion: 'latest',
+    sourceType: 'module',
+    // Options specific to the Babel parser
+    requireConfigFile: false,
+    babelOptions: {
+      plugins: ['@babel/plugin-syntax-import-assertions'],
```
Comment (nchilada) [inline .eslintrc.cjs:35]: This fixes lint for `import ... assert { type: 'json' }`

File: .eslintrc.cjs (hunk 3 above)
Comment (nchilada) [reply .eslintrc.cjs:68]: The bottom line, for developers in this repo: whenever we touch a `*.js`, we'll either need to convert it to ESM in order to lint successfully (and rename it as `*.mjs` if it's in a non-module package, in order to satisfy Node.js), or we'll need to rename it as `*.cjs` (e.g. the `tailwind.config.cjs` for the recalcitrant Tailwind config that @joostdecock mentioned somewhere).

File: .eslintrc.cjs [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    // Additional globals for JavaScript files that happen to be CommonJS.
+    // Only allowed in *.cjs files, not *.js files, because we probably want to move towards
+    // a `"type": "module"` future where any CommonJS files would have to have .cjs extensions.
+    {
+      files: ['**/*.cjs'],
+      env: {
+        commonjs: true,
+      },
+    },
+    // Additional globals for JavaScript files that happen to contain Mocha tests
```
Comment (nchilada) [inline .eslintrc.cjs:75]: `context`, `describe`, `it`

File: .eslintrc.cjs [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        node: true,
+      },
+      rules: {
+        'no-console': 'off',
+      },
+    },
+    {
+      files: frontendFiles,
+      excludedFiles: nodeFiles,
+      extends: ['next/core-web-vitals'],
```
Comment (nchilada) [inline .eslintrc.cjs:53]: This adds support for JSX syntax and also adds a bunch of React- and Next-specific rules

File: .eslintrc.cjs [hunk 8]
```diff
@@ -0,0 +1,114 @@
+const nodeFiles = [
+  '**/build.dflt.{js,mjs,cjs}',
+  '**/build.{js,mjs,cjs}',
+  '**/config/**',
+  '**/prebuild.{js,mjs,cjs}',
+  '**/prebuild/**',
+  '**/scripts/**',
+  'packages/new-design/lib/**',
+  'sites/backend/**',
```
Comment (nchilada) [inline .eslintrc.cjs:9]: Note that when using a single centralized `.eslintrc`, we sometimes have to list specific directories where we want customized lint rules. An alternative would be to add new files, e.g. `sites/backend/.eslintrc`, that override parts (or all) of `./.eslintrc`.

**Review by joostdecock — APPROVED** [review]
:crossed_fingers:

**Comment by nchilada** [comment]
@joostdecock thanks for merging this! I think I was playing with a couple more improvements offline, so I will refresh my memory tomorrow and create a followup PR if applicable

**Pull request merged**